### PR TITLE
Refactoring of Scheduling Distributed Analytical Tasks

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java [hunk 1]
```diff
@@ -8,29 +8,32 @@
...
  *
- * @param <E> the type of entities processed by this task
+ * @param <E> the entity type to be processed by this task
  */
-public interface AnalyticalTask<E> {
+public interface AnalyticalTask<E extends BaseEntity<?>> {
 
     /**
-     * The scheduler which is used to schedules and executes this task for each of its entities.
+     * Returns the class of entities to be processed by this task.
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java]: Could be removed as its the same as the `@return` line

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticsScheduler.java [hunk 2]
```diff
@@ -23,22 +24,34 @@
  * <p>
  * This is a two stage process, {@link #scheduleBatches(Consumer)} emits a list of JSON specification which can
  * be used as descriptions in {@link sirius.biz.cluster.work.DistributedTasks}. Once being executed, the description
- * is forwarded to {@link #collectBatch(JSONObject, Consumer)} which collects and emits all entities of the specified
- * batch to the effective executor of the analyical tasks.
- *
- * @param <E> the type of entities being scheduled
+ * is forwarded to {@link #executeBatch(JSONObject, LocalDate)} which collects all entities of the specified
+ * batch. In most implementations the scheduler will then pick all matchign {@link AnalyticalTask analytical tasks} and
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/AnalyticsScheduler.java]:
```suggestion
 * batch. In most implementations the scheduler will then pick all matching {@link AnalyticalTask analytical tasks} and
```

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticsScheduler.java [hunk 3]
```diff
@@ -23,22 +24,34 @@
...
+ * execute them on the entity.
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/AnalyticsScheduler.java]:
```suggestion
 * executes them on the entity.
```

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticsBatchExecutor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Provides a base implementation for executing batches of tasks generated by an {@link AnalyticsScheduler}.
+ * <p>
+ * This reads the scheduler name and date from each given task description and instructs the appropriate scheduler via
+ * {@link AnalyticsScheduler#executeBatch(JSONObject, LocalDate)} to execute the specified batch.
+ */
+public abstract class AnalyticsBatchExecutor extends DistributedTaskExecutor {
+
+    @Part
+    protected GlobalContext ctx;
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/AnalyticsBatchExecutor.java]: Would prefer the field name `globalContext`

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticsSchedulerExecutor.java [hunk 5]
```diff
@@ -22,10 +22,10 @@
...
- * {@link AnalyticsScheduler#scheduleBatches(Consumer)} to compute all batches to be executed. The emitted
- * are then submitted via {@link DistributedTasks} to be executed by an {@link AnalyticalTaskExecutor}.
+ * {@link AnalyticsScheduler#scheduleBatches(Consumer)} to compute all batches to be executed. The emitted batches
+ * are then submitted via {@link DistributedTasks} to be executed by the appropriate {@link AnalyticsBatchExecutor}.
  */
-abstract class AnalyticsSchedulerExecutor extends DistributedTaskExecutor {
+public abstract class AnalyticsSchedulerExecutor extends DistributedTaskExecutor {
 
     @Part
     protected GlobalContext ctx;
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/AnalyticsSchedulerExecutor.java]: Would prefer the field name `globalContext`

File: src/main/java/sirius/biz/analytics/scheduler/BaseAnalyticalTaskScheduler.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.function.Consumer;
+
+/**
+ * Provides a base implementation which schedules all entities matching a given set of tasks (defined by a
+ * subclass of {@link AnalyticalTask}) using a specified batch size.
+ */
+abstract class BaseAnalyticalTaskScheduler<B extends BaseEntity<?>> implements AnalyticsScheduler {
+
+    @Part
+    protected GlobalContext context;
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/BaseAnalyticalTaskScheduler.java]: Would prefer the field name `globalContext`

File: src/main/java/sirius/biz/analytics/scheduler/SQLEntityBatchEmitter.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                                     .where(OMA.FILTERS.gt(SQLEntity.ID, lastLimit.get()));
+            if (queryExtender != null) {
+                queryExtender.accept(query);
+            }
+
+            query.orderAsc(SQLEntity.ID).limit(batchSize).iterateAll(e -> {
+                nextLimit.set(e.getId());
+            });
+
+            JSONObject batch = new JSONObject();
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/scheduler/SQLEntityBatchEmitter.java]: Missing TYPE?

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java (hunk 1 above)
Comment (andyHa) [reply src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java]: https://www.oracle.com/technetwork/java/javase/documentation/index-137868.html#@return nope

> include it for all other methods, even if its content is entirely redundant with the method description. Having an explicit @return tag makes it easier for someone to find the return value quickly

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticalEngine.java [hunk 8]
```diff
@@ -102,32 +86,52 @@ public void runTimer() throws Exception {
...
+        if (!scheduler.useBestEffortScheduling()) {
+            return true;
+        }
+
+        return queueCache.computeIfAbsent(cluster.getQueueName(scheduler.getExecutorForScheduling()),
+                                          this::checkIfQueueIsEmpty);
+    }
+
+    private Boolean checkIfQueueIsEmpty(String queueName) {
+        return cluster.getQueueLength(queueName) == 0;
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/biz/analytics/scheduler/AnalyticalEngine.java]: to make it more readable^^

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java (hunk 1 above)
Comment (sabieber) [reply src/main/java/sirius/biz/analytics/scheduler/AnalyticalTask.java]: Yeah I mean the first line could be removed. Like this:
```
/**
* @return the class of entities to be processed by this task
*/

File: src/main/java/sirius/biz/analytics/scheduler/AnalyticalEngine.java [hunk 9]
```diff
@@ -102,32 +86,52 @@ public void runTimer() throws Exception {
...
+     * @param scheduler  the scheduler to check
+     * @param queueCache the cached state of the queues being used by the schedulers
+     * @return <tt>true</tt> if the scheduler should run, <tt>false</tt> otherwise
+     */
+    private boolean shouldSchedule(AnalyticsScheduler scheduler, Map<String, Boolean> queueCache) {
+        if (!scheduler.useBestEffortScheduling()) {
+            return true;
+        }
+
+        return queueCache.computeIfAbsent(cluster.getQueueName(scheduler.getExecutorForScheduling()),
```
[the remark this answers is not in the document]
Comment (andyHa) [reply src/main/java/sirius/biz/analytics/scheduler/AnalyticalEngine.java]: not perfect but it looks way better than jamming all this in a giant foreach loop ;-) (been there done that)

**Pull request merged**